Tools: Annotate parameter files with up-to-date documentation

amilcarlucas opened the pull request:
This script fetches on-line ArduPilot documentation (if not cached) and adds it to the specified file or to all *.param and *.parm files in the specified directory.

1. Checks if a local cache of the XML file exists in the target directory or on the directory of the target file:
 - If it does, the script loads the file content.
 - If it doesn't, the script sends a GET request to the URL to fetch the XML data for the requested vehicle type.
2. Parses the XML data and creates a dictionary of parameter documentation.
3. DELETES all comments that start at the beginning of a line
4. Adds the parameter documentation to the target file or to all *.param,*.parm files in the target directory.

Supports AP_Periph, AntennaTracker, ArduCopter, ArduPlane, ArduSub, Blimp, Heli, Rover and SITL vehicle types
Supports both Mission Planner and MAVProxy file formats
Supports sorting the parameters
Has unit tests with 88% coverage

AP_FLAKE8_CLEAN

Author: Amilcar do Carmo Lucas, IAV GmbH

IamPete1 commented:
Can you upload a example output?

One issue I foresee is that param files don't contain version information, so you get the latest descriptions for a parameter file that may be for a older version.

We could get the webtools to do a similar thing to the various parameter files they generate.

amilcarlucas commented:
Yes, [examples for multiple versions are available here](https://discuss.ardupilot.org/uploads/short-url/j5A0TRSD8MpfKp9B9WCW5n95KfJ.zip) There you can see how the tool was used to generate documentation for different versions.

peterbarker commented on Tools/autotest/param_metadata/param_docs_update.py:
```diff
@@ -0,0 +1,362 @@
+If it is, it prefixes the line with the corresponding value from the dictionary. If it is not, it simply adds the line to the new lines.
+
+Amilcar do Carmo Lucas, IAV GmbH
+"""
```
Should be marked `AP_FLAKE8_CLEAN`

peterbarker commented on Tools/autotest/param_metadata/param_docs_update.py:
```diff
@@ -0,0 +1,362 @@
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Fetches on-line ArduPilot parameter documentation and adds it to the specified file or to all *.param and *.parm files in the specified directory.')
+    parser.add_argument('target', nargs='?', default='.', help='The target file or directory. Defaults to the current directory.')
+    choices = ['none', 'missionplanner', 'mavproxy']
+    parser.add_argument('-s', '--sort', default='none', choices=choices, help='Sort the parameters in the file. Defaults to not sorting.')
```
```suggestion
    parser.add_argument(
        '-s', '--sort',
        choices=['none', 'missionplanner', 'mavproxy'],
        default='none',
        help='Sort the parameters in the file. Defaults to not sorting.',
    )
```

etc

peterbarker reviewed (changes requested):
Interesting.

You could potentially make the vehicle-type parameter optional, allowing the tool to search the param file for parameters which are strictly only present on a particular vehicle type (e.g. `Q_ENABLE`)

I'm wondering whether this should be moved out of `autotest`, perhaps back into `Tools/scripts`.  And perhaps a name change to `annotate_parm_file.py`?

Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?

Doesn't actually do what the help says it does, in that if `apm.pdef.xml` already exists it doesn't fetch it.  Perhaps fetch by default (over-writing), but add a `--cached` option.

Consider adding some comments at the top explaining how the file was annotated - i.e. annotated on such-and-such a date from an `apm.pdef.xml` dated such-and-such with checksum such-and-such.  And which vehicle it was assuming when it annotated!

Partial review only, sorry.

amilcarlucas commented:
> Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?

The documentation can be improved :) It already supports directory or file to be passed in

peterbarker commented:
On Mon, 22 Jan 2024, Amilcar Lucas wrote:

>       Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?
> 
> The documentation can be improved :) It already supports directory or file to be passed in

Right, I'm suggesting removing the magic "take all files in the current
directory" thing.  *Require* an argument there.  And perhaps make that a 
file only.  Or make it one-or-many files to be valid.  Just not the magic 
default behaviour.

amilcarlucas commented:
>You could potentially make the vehicle-type parameter optional, allowing the tool to search the param file for parameters which are strictly only present on a particular vehicle type (e.g. Q_ENABLE)

For my use case (intermediate parameter files with just a couple of parameters) it would not work.

> I'm wondering whether this should be moved out of autotest, perhaps back into Tools/scripts. And perhaps a name change to annotate_parm_file.py?

I moved it to Tools/scripts

> Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?

I have directories with over 40 files and I want to do them all at once :)

> Doesn't actually do what the help says it does, in that if apm.pdef.xml already exists it doesn't fetch it. Perhaps fetch by default (over-writing), but add a --cached option.

Fixed the documentation

> Consider adding some comments at the top explaining how the file was annotated - i.e. annotated on such-and-such a date from an apm.pdef.xml dated such-and-such with checksum such-and-such. And which vehicle it was assuming when it annotated!

Very good idea, but that information needs to be present in the `apm.pdef.xml` file in order for it to work like it should.
I leave that for a for a `param_parse.py` future PR

peterbarker commented:
> > Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?
> 
> The documentation can be improved :) It already supports directory or file to be passed in

Sorry, I'm quite aware of that.  Principle of least surprise says they *must* be supplied.

peterbarker commented:
> > Also, annotating an entire directory is kind of blunt; perhaps require a specific file be passed in?
> 
> I have directories with over 40 files and I want to do them all at once :)

Sure.  `./Tools/scripts/annotate_params.py .`

amilcarlucas commented:
Thanks Peter it is clear now. :)

amilcarlucas commented:
@peterbarker @khancyr did I add the unit test to the github action correctly?

amilcarlucas commented:
@IamPete1 ist the output format satisfactory?

peterbarker commented on Tools/scripts/annotate_params_unittest.py:
```diff
@@ -0,0 +1,536 @@
+#!/usr/bin/python3
```
This one should go elsewhere - perhaps `Tools/autotest/unittest`?

amilcarlucas replied:
Done

peterbarker reviewed (changes requested):
@khancyr is bringing up an environment in github actions like this the appropriate way to CI this script, or should be it woven into some other place?

tridge commented:
@amilcarlucas need to ping @IamPete1 on code_review channel to discuss

khancyr commented on .github/workflows/test_annotate_params.yml:
```diff
@@ -0,0 +1,28 @@
+  build:
+    runs-on: ubuntu-22.04
+    container:
+      image: ardupilot/ardupilot-dev-clang:v0.1.3
```
dev-base

amilcarlucas replied:
done

khancyr commented on .github/workflows/test_annotate_params.yml:
```diff
@@ -0,0 +1,28 @@
+    runs-on: ubuntu-22.04
+    container:
+      image: ardupilot/ardupilot-dev-clang:v0.1.3
+      options: --user 1001
```
needed ?

amilcarlucas replied:
Done, I think

amilcarlucas replied:
done

khancyr commented on .github/workflows/test_annotate_params.yml:
```diff
@@ -0,0 +1,28 @@
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Set up Python
```
we don't need the python setup nor Install dependencies as this is already done by the base image

amilcarlucas replied:
Done, I think

tridge commented on Tools/scripts/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Automation Scripts"""
```
why this? this is not a python module
